features/cellular/TARGET_GENERIC_MODEM/generic_modem_driver/ReferenceCellularInterface.cpp

```diff
@@ -0,0 +1,794 @@
+            tr_debug("Registered for circuit switched service with CSFB not preferred");
+            break;
+        default:
+            tr_debug("Unknown circuit switched service registration status. %u", status);
+            break;
+    }
+
+    dev_info->reg_status_csd = static_cast<nwk_registration_status_csd>(status);
+}
+
+static void set_nwk_reg_status_psd(unsigned int status)
+{
+    switch (status) {
+        case PSD_NOT_REGISTERED_NOT_SEARCHING:
+        case PSD_NOT_REGISTERED_SEARCHING:
+            break;
+        case PSD_REGISTERED:
+        case PSD_REGISTERED_ROAMING:
+            tr_debug("Registered for packet switched service");
+            break;
+        case PSD_REGISTRATION_DENIED:
+            tr_debug("Packet switched service denied");
+            break;
+        case PSD_UNKNOWN_COVERAGE:
+            tr_debug("Out of packet switched service coverage");
+            break;
+        case PSD_EMERGENCY_SERVICES_ONLY:
+            tr_debug("Limited access for packet switched service. Emergency use only.");
+            break;
+        default:
+            tr_debug("Unknown packet switched service registration status. %u", status);
+            break;
+    }
+
+    dev_info->reg_status_psd = static_cast<nwk_registration_status_psd>(status);
+}
+
+static bool is_registered_csd()
+{
+  return (dev_info->reg_status_csd == CSD_REGISTERED) ||
+          (dev_info->reg_status_csd == CSD_REGISTERED_ROAMING) ||
+          (dev_info->reg_status_csd == CSD_CSFB_NOT_PREFERRED);
+}
+
+static bool is_registered_psd()
+{
+    return (dev_info->reg_status_psd == PSD_REGISTERED) ||
+            (dev_info->reg_status_psd == PSD_REGISTERED_ROAMING);
+}
+
+ReferenceCellularInterface::ReferenceCellularInterface(bool debugOn, PinName tx, PinName rx, int baud)
+{
+    _new_pin = NULL;
+    _pin = NULL;
+    _at = NULL;
+    _dcd = NULL;
+    _apn = "internet";
+    _uname = NULL;
+    _pwd = NULL;
+    _debug_trace_on = false;
+
+    // Set up File Handle
+    _fh = new BufferedSerial(tx, rx, BAUD_RATE);
+
+    if (debugOn) {
+         _debug_trace_on = true;
+     }
+
+    dev_info = new device_info;
+    dev_info->reg_status_csd = CSD_NOT_REGISTERED_NOT_SEARCHING;
+    dev_info->reg_status_psd = PSD_NOT_REGISTERED_NOT_SEARCHING;
+    dev_info->ppp_connection_up = false;
+
+}
+
+ReferenceCellularInterface::~ReferenceCellularInterface()
+{
+    delete _fh;
+    delete _at;
+    delete _dcd;
+    delete dev_info;
+}
+
+void ReferenceCellularInterface::connection_status_cb(void (*fptr)(nsapi_error_t))
+{
+    callback_fptr = fptr;
+}
+
+/**
+ * Public API. Sets up the flag for the driver to enable or disable SIM pin check
+ * at the next boot.
+ */
+void ReferenceCellularInterface::check_sim_pin(bool check)
+{
+    set_sim_pin_check_request = check;
+}
+
+/**
+ * Public API. Sets up the flag for the driver to change pin code for SIM card
+ */
+void ReferenceCellularInterface::change_sim_pin(const char *new_pin)
```

Comment:
Why is there `set_sim_pin` and `change_sim_pin`? Can't we check if the pin has changed for the user? This function seems unnecessary.